src/solvers/flattening/arrays.cpp

```diff
@@ -185,9 +189,10 @@ void arrayst::collect_arrays(const exprt &a)
   }
   else if(a.id()==ID_member)
   {
-    if(to_member_expr(a).struct_op().id()!=ID_symbol)
-      throw
-        "unexpected array expression: member with `"+a.op0().id_string()+"'";
+    DATA_INVARIANT(
+      to_member_expr(a).struct_op().id()==ID_symbol,
+      ("unexpected array expression: member with `"+
+       a.op0().id_string()+"'").c_str());
```

Comment:
It would also be good to make this lazy for cases like this which build an error string. How about something like:
```
#define INVARIANT(CONDITION, REASON) \
  do { \
    if(!(CONDITION)) \
    { \
      std::string reason = (REASON); \
      invariant_failed(__FILE__, __FUNCTION__, __LINE__, reason); \
    } \
  } while(0);
```